### PR TITLE
Add UI for Schedule Helper

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

Adds the ability to use the Schedule Helper.

No Overlapping events
No Cross day events
No dragging over events
No resizing over events
Click to Delete

CleanShot 2022-08-11 at 15 40 12

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/data/schedule.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  });
+
+export const deleteSchedule = (hass: HomeAssistant, id: string) =>
+  hass.callWS({
+    type: "schedule/delete",
+    schedule_id: id,
+  });
+
+export const getScheduleTime = (date: Date): string =>
+  `${("0" + date.getHours()).slice(-2)}:${("0" + date.getMinutes()).slice(-2)}`;
```
Comment (balloob) [inline src/data/schedule.ts:65]: Use a date formatting function

File: src/data/schedule.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import { HomeAssistant } from "../types";
+
+export const weekdays = [
+  "sunday",
```
Comment (balloob) [inline src/data/schedule.ts:4]: Can we start on Monday?

File: src/panels/config/helpers/forms/ha-schedule-form.ts [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+class HaScheduleForm extends LitElement {
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @property() public new?: boolean;
+
+  @state() private _name!: string;
+
+  @state() private _icon!: string;
+
+  // @ts-ignore
```
Comment (balloob) [inline src/panels/config/helpers/forms/ha-schedule-form.ts:58]: What are you ignoring

**Review by balloob — DISMISSED** [review]

File: src/data/schedule.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  id: string;
+  name: string;
+  icon?: string;
+  monday?: ScheduleDay[];
+  tuesday?: ScheduleDay[];
+  wednesday?: ScheduleDay[];
+  thursday?: ScheduleDay[];
+  friday?: ScheduleDay[];
+  saturday?: ScheduleDay[];
+  sunday?: ScheduleDay[];
```
Comment (steverep) [inline src/data/schedule.ts:28]:
```suggestion
  [day: typeof weekdays[number]]?: ScheduleDay[];
```

File: src/data/schedule.ts (hunk 2 above)
Comment (zsarnett) [reply src/data/schedule.ts:4]: Not for this. JS starts on Sunday

File: src/panels/config/helpers/forms/ha-schedule-form.ts (hunk 3 above)
Comment (zsarnett) [reply src/panels/config/helpers/forms/ha-schedule-form.ts:58]: Unused variable

Comment (steverep) [reply src/panels/config/helpers/forms/ha-schedule-form.ts:58]: Wondering the same.  Does the change I suggested help?  Should force TS to understand `this[_${day}]` better.

File: src/data/schedule.ts (hunk 2 above)
Comment (Misiu) [reply src/data/schedule.ts:4]: Ideally, the calendar should start on Monday, just like shown here: https://github.com/home-assistant/core/pull/76566#issuecomment-1211673836

Comment (matthiasdebaat) [reply src/data/schedule.ts:4]: Would be nice to keep it consistent

Comment (balloob) [reply src/data/schedule.ts:4]: @zsarnett right, for this array we track JS indexes, but that doesn't mean our UI can't start on Monday

Comment (zsarnett) [reply src/data/schedule.ts:4]: It starts based on your locale

File: src/panels/config/helpers/forms/ha-schedule-form.ts (hunk 3 above)
Comment (zsarnett) [reply src/panels/config/helpers/forms/ha-schedule-form.ts:58]: No your suggestion doesn't modify the element or class just the Schedule type

File: src/data/schedule.ts (hunk 4 above)
Comment (zsarnett) [reply src/data/schedule.ts:28]: Not sure if this is better... just more confusing to read later maybe haha

I added it with the correct syntax

File: src/panels/config/helpers/forms/ha-schedule-form.ts (hunk 3 above)
Comment (zsarnett) [reply src/panels/config/helpers/forms/ha-schedule-form.ts]: It does yes. Thanks

File: src/data/schedule.ts (hunk 2 above)
Comment (zsarnett) [reply src/data/schedule.ts:4]: > Would be nice to keep it consistent

Consistent? Thats node red not HA

File: src/data/schedule.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+type ScheduleDays = { [K in typeof weekdays[number]]?: ScheduleDay[] };
+
+export interface Schedule extends ScheduleDays {
+  id: string;
+  name: string;
+  icon?: string;
+}
+
+export interface ScheduleMutableParams {
```
Comment (balloob) [inline src/data/schedule.ts:26]: Is icon not mutable?

Comment (zsarnett) [reply src/data/schedule.ts:27]: I guess everything is 🤷

**Review by balloob — DISMISSED** [review]

File: src/data/schedule.ts (hunk 5 above)
Comment (balloob) [reply src/data/schedule.ts:26]: id is not :)

**Review by balloob — APPROVED** [review]

**Pull request merged**